### PR TITLE
Added CCAM (Convolutional Coordinate Attention Module), enhances the model's ability to focus on important features by applying attention mechanisms in both channel and spatial dimensions.

### PR DESCRIPTION
**Pull request by Anawaert**
Dear Ultralytics, recently while researching attention mechanisms and the YOLO11 network architecture, I noticed that CBAM might have a more lightweight alternative. I combined the Channel Attention Mechanism (CAM) with the Coordinate Attention Mechanism (CA) and proposed a new module called the Convolutional-Coordinate Attention Mechanism (CCAM). After integrating CCAM between the Neck and Head of the network, we observed noticeable improvements on our internal research dataset, without significant increases in parameter count or computational complexity.  
Below is a comparison table of evaluation metrics, and YOLO11m-seg is used for editing and test.

| CCAM | $P$ | $R$ | ${mAP}_{50}$ | ${mAP}_{50:95}$ | Parameters | GFLOPs |
|----------|-------|-------|---------------------|------------------------|-----------------|--------------|
|    √      | $88.5%$ | $77.9%$ | $87.0%$ | $66.7%$ | $23.6$ | $124.7$               |
|    ×     | $85.3%$ | $77.1%$ | $86.0%$ | $64.6%$ | $22.3$ | $123.9$               |

Hpoing that CCAM can be considered as an optional block or convolution layer within Ultralytics, I would appreciate it if you can evaluate it and merge this branch!

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Introduces a new Convolutional Coordinate Attention Module (CCAM) to enhance feature extraction via combined channel and spatial attention, and wires it into the Ultralytics model registry for easy use. 🔧🧠

### 📊 Key Changes
- Added CCAM module to nn.modules.block with forward and init implementations.
- Integrated ChannelAttention dependency and supporting ops (pooling, Conv, Hardswish).
- Registered CCAM in module exports (__all__), modules/__init__.py, and nn/tasks.py parse_model so it can be used in model YAMLs.
- Ensures deterministic pooling by temporarily moving tensors to CPU in CCAM’s coordinate attention path.

### 🎯 Purpose & Impact
- Better feature representation: Combines Channel Attention and Coordinate Attention to help models focus on important channels and spatial regions. 📈
- Easy adoption: CCAM is now a first-class module selectable in architectures and custom YAMLs. ⚙️
- Potential accuracy gains: May improve detection/segmentation performance with minimal architectural changes. 🎯
- Trade-offs: Slight extra compute/latency due to attention operations and CPU transfer in pooling; users should benchmark for their use case. ⏱️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Anawaert, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review process. An Ultralytics engineer will also review your contribution and provide further assistance soon.

To ensure your Convolutional Coordinate Attention Module (CCAM) integration proceeds smoothly, please review the following checklist:

- ✅ **Define a Purpose**: Your PR description is detailed—thank you! If you have any related [issues](https://github.com/ultralytics/ultralytics/issues), please link them for additional context.
- ✅ **Synchronize with Source**: Confirm your PR is up to date with the `ultralytics/ultralytics` `main` branch. If not, update by clicking 'Update branch' or running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Make sure all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address them.
- ✅ **Update Documentation**: Update [documentation](https://docs.ultralytics.com/) to reflect your new or modified modules, especially for CCAM usage.
- ✅ **Add Tests**: If applicable, include or update tests to cover CCAM integration and verify all tests pass.
- ✅ **Sign the CLA**: If this is your first PR to Ultralytics, please sign our [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/) by commenting "I have read the CLA Document and I sign the CLA" below.
- ✅ **Minimize Changes**: Please ensure your changes are limited to the **minimum** necessary for CCAM integration. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

If your PR addresses a bug and you haven’t already, please provide a **minimum reproducible example (MRE)** to help us validate and test your changes effectively.

For more guidance, see our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If you have any questions, feel free to comment below. Thank you for helping improve Ultralytics! 🚀🧠

**Comment by Anawaert** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
@Anawaert thank you for your work here! @Laughing-q can you take a look?

**Comment by Laughing-q** [comment]
@Anawaert Thank for sharing the insights here! I'll test it out and update here if I found anything!

File: ultralytics/nn/modules/block.py [hunk 1]
```diff
@@ -1964,3 +1965,72 @@ def forward(self, x, vp):
...
+        # Initialize the channel attention.
+        self.channel_attention = ChannelAttention(c1)
+
+        # Initialize the coordinate attention.
+        self.ca_avg_pool_h = nn.AdaptiveAvgPool2d((None, 1))
+        self.ca_avg_pool_w = nn.AdaptiveAvgPool2d((1, None))
+        mip = max(8, c1 // reduction)
+        self.conv1 = nn.Conv2d(in_channels=c1, out_channels=mip, kernel_size=1, stride=1, padding=0)
+        self.BN = nn.BatchNorm2d(mip)
+        self.act = nn.Hardswish()
```
Comment (Laughing-q) [inline ultralytics/nn/modules/block.py:1998]: @Anawaert FYI by using `conv+bn+act` routine I would suggest using our internal `Conv` module which is a wrapper of the `conv+bn+act` combination. https://github.com/ultralytics/ultralytics/blob/20196b57d0285856d2b9e9ef095beb39d0643c8e/ultralytics/nn/modules/conv.py#L37-L79

**Comment by Anawaert** [comment]
> @Anawaert Could you please also share where you inject this module into model layers?

Of course I can! The following image shows that how did I inject this module into YOLO11 model layers, between Neck and Head.
CCAM_location

File: ultralytics/nn/modules/block.py (hunk 1 above)
Comment (Anawaert) [reply ultralytics/nn/modules/block.py]: Okay, I just pushed a new commit to the CCAM_devel branch, using internal `Conv` block instead of `conv+bn_act` method, hoping that you can test it successfully, thx.

Comment (Laughing-q) [reply ultralytics/nn/modules/block.py]: @Anawaert Awesome! Thanks!

**Comment by Laughing-q** [comment]
@Anawaert Thank you!

**Comment by Laughing-q** [comment]
@Anawaert Hey I was testing the new module here and here's a warning appears
```bash
UserWarning: adaptive_avg_pool2d_backward_cuda does not have a det erministic implementation
```
which I believe is related to these two `nn.AdaptiveAvgPool2d` here
https://github.com/ultralytics/ultralytics/blob/CCAM_devel/ultralytics/nn/modules/block.py#L1993-L1994
Ideally we prefer training to be deterministic hence users could always reproduce consistent results. Do you know if there's any solution to fix this, or alternatives here?

**Comment by Anawaert** [comment]
> @Anawaert Hey I was testing the new module here and here's a warning appears
> 
> ```shell
> UserWarning: adaptive_avg_pool2d_backward_cuda does not have a det erministic implementation
> ```
> 
> which I believe is related to these two `nn.AdaptiveAvgPool2d` here https://github.com/ultralytics/ultralytics/blob/CCAM_devel/ultralytics/nn/modules/block.py#L1993-L1994 Ideally we prefer training to be deterministic hence users could always reproduce consistent results. Do you know if there's any solution to fix this, or alternatives here?

@Laughing-q  Yes, I actually noticed this issue earlier during injection testing. Initially, I considered manually implementing an adaptive average pooling layer myself. However, given that `nn.AdaptiveAvgPool` offers better generality and compatibility, I eventually abandoned that idea.  

But just now, it occurred to me that if we move the pooling operation to the CPU, the results could become deterministic. So I made some modifications to the code — I only moved the pooling operation to the CPU, while keeping all other steps exactly the same as before. After some rough testing, no warnings appeared across 30 rounds of training, indicating that this approach is effective. Because the CPU was used, the training time increased slightly. On my dataset, the time per training epoch rose from 1 minute 20 seconds to 1 minute 38 seconds, and the validation time increased from 4 seconds to 5 seconds.

Using non-global adaptive pooling in the current version of PyTorch is indeed quite a headache. Maybe we’ll just have to hope that PyTorch can eventually achieve deterministic implementation of adaptive pooling on CUDA in the future.

For now, the latest code has been pushed to the CCAM_devel branch, and you can test it as well as evaluate it!

**Comment by Anawaert** [comment]
@Laughing-q Hi~, is there anything else I can help with or I need to provide?

**Comment by Laughing-q** [comment]
@Anawaert Hey thanks for the implementation here! However I got slightly lower results here within this attention module compare to our baseline yolo11n on coco:
| Model             | mAP  |
|-------------------|------|
| yolo11n(baseline) | 39.3 |
| yolo11n-ccam      | 39.1 |

Let's leave this PR opened, we might want to revisit it in the future!

**Comment by Anawaert** [comment]
@Laughing-q OMG, it performed a little worse on COCO dataset within yolo11n 😥. I originally thought that it would perform better whether in COCO or my own dataset.

Okay, I'll keep the PR open, and I'll try to do more improvements and experiments in the future. Thanks for your review. 😀

**Comment by Laughing-q** [comment]
@Anawaert Awesome! Thank you!
